mindmeld/active_learning/alp.py

```diff
@@ -0,0 +1,273 @@
+            "test_pattern": self.test_pattern,
+            "load": self.load,
+            "save": self.save,
+            "train_seed_pct": self.train_seed_pct,
+            "n_classifiers": self.n_classifiers,
+            "n_epochs": self.n_epochs,
+            "batch_size": self.batch_size,
+            "training_level": self.training_level,
+            "training_strategies": self.training_strategies,
+            "selection_strategy": self.selection_strategy,
+            "save_sampled_queries": self.save_sampled_queries,
+            "log_usage_pct": self.log_usage_pct,
+            "labeled_logs_pattern": self.labeled_logs_pattern,
+            "unlabled_logs_path": self.unlabeled_logs_path,
+            "output_folder": self.output_folder,
+        }
+
+    def train(self):
+        """Loads the initial data bucket and then trains on every strategy."""
+        logger.info("Creating Output Folder and Saving Params.")
+        self.results_manager.create_experiment_folder(
+            active_learning_params=self._as_dict(),
+            training_strategies=self.training_strategies,
+        )
+        logger.info("Creating Training Data Bucket.")
+        self.init_train_data_bucket = DataBucketFactory.get_data_bucket_for_training(
+            self.app_path,
+            self.load,
+            self.save,
+            self.training_level,
+            self.train_pattern,
+            self.test_pattern,
+            self.train_seed_pct,
+        )
+        logger.info("Starting Training")
+        self._train_all_strategies()
+
+    def select(self):
+        """Selects the next batch of queries to label from a set of log queries."""
+        logger.info("Loading Queries for active learning.")
+        self.init_train_data_bucket = DataBucketFactory.get_data_bucket_for_selection(
+            self.app_path,
+            self.load,
+            self.save,
+            self.training_level,
+            self.train_pattern,
+            self.test_pattern,
+            self.unlabeled_logs_path,
+            self.labeled_logs_pattern,
+            self.log_usage_pct,
+        )
+        logger.info("Starting Selection.")
+        self._train_single_strategy(
+            strategy=self.selection_strategy, training_mode=False
+        )
+        self.results_manager.write_log_selected_queries_json(
+            strategy=self.selection_strategy,
+            queries=self.data_bucket.newly_sampled_queries,
+        )
+
+    def plot(self):
+        """Creates the generated folder and its subfolders if they do not already exist."""
+        plot_manager = PlotManager(self.results_manager.experiment_folder)
+        plot_manager.generate_plots()
+
+    def _train_all_strategies(self):
+        """ Train with all active learning strategies."""
+        for strategy in self.training_strategies:
+            self._train_single_strategy(strategy)
+
+    def _train_single_strategy(self, strategy: str, training_mode: bool = True):
```

Comment:
An alternate suggestion is to change `_train_all_strategies` to `_run_all_strategies` and `_train_single_strategy` to `_run_single_strategy` since we're not really "training" anything. That way having a `training_mode` param will make sense.

Reply:
Good points. Changed "_train_single_strategy" to "_run_strategy" and "training_mode" to "select_mode". Keeping "_train_all_strategies" as this function is specific for training.